File: examples/c/jacobi_1d.c
// Copyright (c) 2026 Bryce M. Westheimer
// SPDX-License-Identifier: BSD-3-Clause

/**
 * @file jacobi_1d.c
 * @brief 1D Jacobi iterative solver with DTL C bindings
 *
 * Solves u''(x) = 0 with boundary conditions u(0) = 1, u(L) = 0.
 * Uses Jacobi iteration with halo exchange via dtl_send / dtl_recv.
 *
 * Demonstrates:
 * - Halo exchange using dtl_send / dtl_recv
 * - Convergence checking via dtl_allreduce with DTL_OP_MAX
 * - Distributed iterative solver pattern
 *
 * Run:
 *   mpirun -np 4 ./c_jacobi_1d
 */

#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <dtl/bindings/c/dtl.h>

#define CHECK_STATUS(status, msg) \
    do { \
        if (!dtl_status_ok(status)) { \
            fprintf(stderr, "%s: %s\n", msg, dtl_status_message(status)); \
            exit(1); \
        } \
    } while(0)

int main(int argc, char** argv) {
    (void)argc;
    (void)argv;

    dtl_context_t ctx = NULL;
    dtl_status status;

    status = dtl_context_create_default(&ctx);
    CHECK_STATUS(status, "Failed to create context");

    dtl_rank_t rank = dtl_context_rank(ctx);
    dtl_rank_t size = dtl_context_size(ctx);

    if (rank == 0) {
        printf("DTL 1D Jacobi Solver (C)\n");
        printf("=========================\n");
        printf("Ranks: %d\n", size);
    }
    dtl_barrier(ctx);

    /* Problem setup */
    const int global_n = 100;       /* Interior points */
    const int max_iter = 10000;
    const double tol = 1e-8;

    /* Partition interior points among ranks */
    int local_n = global_n / size;
    int remainder = global_n % size;
    if (rank < remainder) local_n += 1;

    /* Allocate arrays with halo cells: [left_halo | interior | right_halo] */
    double* u     = (double*)calloc((size_t)(local_n + 2), sizeof(double));
    double* u_new = (double*)calloc((size_t)(local_n + 2), sizeof(double));
    if (!u || !u_new) {
        fprintf(stderr, "Memory allocation failed\n");
        exit(1);
    }

    /* Apply boundary conditions: u(0) = 1.0 on leftmost rank */
    if (rank == 0) {
        u[0] = 1.0;
        u_new[0] = 1.0;
    }
    /* u(L) = 0.0 on rightmost rank's right boundary (already 0) */

    const int halo_tag = 10;

    if (rank == 0) {
        printf("Grid: %d interior points\n", global_n);
        printf("BCs: u(0)=1, u(L)=0\n");
        printf("Tolerance: %.0e\n\n", tol);
    }
    dtl_barrier(ctx);

    int iter = 0;
    double global_diff = 0.0;

    for (iter = 0; iter < max_iter; ++iter) {
        /* --- Halo exchange --- */
        /* Even ranks send right, then left; odd ranks receive first */
        /* Send right boundary to right neighbor */
        if (rank < size - 1) {
            status = dtl_send(ctx, &u[local_n], 1, DTL_DTYPE_FLOAT64,
                              rank + 1, halo_tag);
            CHECK_STATUS(status, "Send right failed");
        }
        if (rank > 0) {
            status = dtl_recv(ctx, &u[0], 1, DTL_DTYPE_FLOAT64,
                              rank - 1, halo_tag);
            CHECK_STATUS(status, "Recv left failed");
        }

        /* Send left boundary to left neighbor */
        if (rank > 0) {
            status = dtl_send(ctx, &u[1], 1, DTL_DTYPE_FLOAT64,
                              rank - 1, halo_tag + 1);
            CHECK_STATUS(status, "Send left failed");
        }
        if (rank < size - 1) {
            status = dtl_recv(ctx, &u[local_n + 1], 1, DTL_DTYPE_FLOAT64,
                              rank + 1, halo_tag + 1);
            CHECK_STATUS(status, "Recv right failed");
        }

        /* --- Jacobi update --- */
        double local_diff = 0.0;
        for (int i = 1; i <= local_n; ++i) {
            u_new[i] = 0.5 * (u[i - 1] + u[i + 1]);
            double d = fabs(u_new[i] - u[i]);
            if (d > local_diff) local_diff = d;
        }

        /* Copy new to old */
        for (int i = 1; i <= local_n; ++i) {
            u[i] = u_new[i];
        }

        /* Check convergence: global max diff */
        status = dtl_allreduce(ctx, &local_diff, &global_diff, 1,
                               DTL_DTYPE_FLOAT64, DTL_OP_MAX);
        CHECK_STATUS(status, "Allreduce max failed");

        if (global_diff < tol) {
            break;
        }
    }

    dtl_barrier(ctx);

    if (rank == 0) {
        printf("Converged after %d iterations\n", iter);
        printf("Final max diff: %.4e\n\n", global_diff);
    }

    /* Print solution samples */
    for (dtl_rank_t r = 0; r < size; ++r) {
        if (rank == r) {
            printf("  Rank %d: u[first]=%.6f, u[last]=%.6f\n",
                   rank, u[1], u[local_n]);
        }
        dtl_barrier(ctx);
    }

    if (rank == 0) {
        printf("\nExpected: linear from 1.0 to 0.0\n");
        printf("%s: Solver %s\n",
               global_diff < tol ? "SUCCESS" : "FAILURE",
               global_diff < tol ? "converged" : "did not converge");
    }

    free(u);
    free(u_new);
    dtl_context_destroy(ctx);
    return 0;
}

File: examples/fortran/jacobi_1d.f90
! Copyright (c) 2026 Bryce M. Westheimer
! SPDX-License-Identifier: BSD-3-Clause

!> @file jacobi_1d.f90
!> @brief 1D Jacobi iterative solver with DTL Fortran bindings
!>
!> Solves u''(x) = 0 with boundary conditions u(0) = 1, u(L) = 0.
!> Uses Jacobi iteration with halo exchange via dtl_send / dtl_recv.
!>
!> Demonstrates:
!> - Halo exchange using dtl_send / dtl_recv
!> - Convergence checking via dtl_allreduce with DTL_OP_MAX
!> - Distributed iterative solver pattern
!>
!> Run:
!>   mpirun -np 4 ./fortran_jacobi_1d

program jacobi_1d
    use, intrinsic :: iso_c_binding
    use dtl_fortran
    implicit none

    type(c_ptr) :: ctx
    integer(c_int32_t) :: status
    integer(c_int32_t) :: rank, num_ranks
    integer :: global_n, max_iter, local_n, remainder_n
    real(c_double) :: tol
    real(c_double), allocatable, target :: u(:), u_new(:)
    real(c_double), target :: local_diff, global_diff
    integer :: iter, i
    integer :: halo_tag

    ! Create context
    status = dtl_context_create_default(ctx)
    if (.not. is_success(status)) then
        print '(A)', 'ERROR: Failed to create context'
        stop 1
    end if

    rank = dtl_context_rank(ctx)
    num_ranks = dtl_context_size(ctx)

    if (rank == 0) then
        print '(A)', 'DTL 1D Jacobi Solver (Fortran)'
        print '(A)', '==============================='
        print '(A,I0)', 'Ranks: ', num_ranks
    end if
    status = dtl_barrier(ctx)

    ! Problem setup
    global_n = 100       ! Interior points
    max_iter = 10000
    tol = 1.0e-8_c_double
    halo_tag = 10

    ! Partition interior points among ranks
    local_n = global_n / num_ranks
    remainder_n = mod(global_n, num_ranks)
    if (rank < remainder_n) local_n = local_n + 1

    ! Allocate arrays with halo cells: index 0 = left halo, local_n+1 = right halo
    allocate(u(0:local_n + 1))
    allocate(u_new(0:local_n + 1))
    u = 0.0_c_double
    u_new = 0.0_c_double

    ! Boundary conditions: u(0) = 1.0 on leftmost rank
    if (rank == 0) then
        u(0) = 1.0_c_double
        u_new(0) = 1.0_c_double
    end if
    ! u(L) = 0.0 on rightmost rank (already zero)

    if (rank == 0) then
        print '(A,I0,A)', 'Grid: ', global_n, ' interior points'
        print '(A)', 'BCs: u(0)=1, u(L)=0'
        print '(A,ES8.0)', 'Tolerance: ', tol
        print '(A)', ''
    end if
    status = dtl_barrier(ctx)

    global_diff = 0.0_c_double

    do iter = 1, max_iter
        ! --- Halo exchange ---
        ! Send right boundary to right neighbor
        if (rank < num_ranks - 1) then
            status = dtl_send(ctx, c_loc(u(local_n)), 1_c_int64_t, &
                              DTL_DTYPE_FLOAT64, rank + 1, halo_tag)
        end if
        if (rank > 0) then
            status = dtl_recv(ctx, c_loc(u(0)), 1_c_int64_t, &
                              DTL_DTYPE_FLOAT64, rank - 1, halo_tag)
        end if

        ! Send left boundary to left neighbor
        if (rank > 0) then
            status = dtl_send(ctx, c_loc(u(1)), 1_c_int64_t, &
                              DTL_DTYPE_FLOAT64, rank - 1, halo_tag + 1)
        end if
        if (rank < num_ranks - 1) then
            status = dtl_recv(ctx, c_loc(u(local_n + 1)), 1_c_int64_t, &
                              DTL_DTYPE_FLOAT64, rank + 1, halo_tag + 1)
        end if

        ! --- Jacobi update ---
        local_diff = 0.0_c_double
        do i = 1, local_n
            u_new(i) = 0.5_c_double * (u(i - 1) + u(i + 1))
            local_diff = max(local_diff, abs(u_new(i) - u(i)))
        end do

        ! Copy new to old
        u(1:local_n) = u_new(1:local_n)

        ! Check convergence: global max diff
        status = dtl_allreduce(ctx, c_loc(local_diff), c_loc(global_diff), &
                               1_c_int64_t, DTL_DTYPE_FLOAT64, DTL_OP_MAX)

        if (global_diff < tol) exit
    end do

    status = dtl_barrier(ctx)

    if (rank == 0) then
        print '(A,I0,A)', 'Converged after ', iter - 1, ' iterations'
        print '(A,ES12.4)', 'Final max diff: ', global_diff
        print '(A)', ''
    end if

    ! Print solution samples
    block
        integer(c_int32_t) :: r
        do r = 0, num_ranks - 1
            if (rank == r) then
                print '(A,I0,A,F10.6,A,F10.6)', &
                    '  Rank ', rank, ': u[first]=', u(1), ', u[last]=', u(local_n)
            end if
            status = dtl_barrier(ctx)
        end do
    end block

    if (rank == 0) then
        print '(A)', ''
        print '(A)', 'Expected: linear from 1.0 to 0.0'
        if (global_diff < tol) then
            print '(A)', 'SUCCESS: Solver converged'
        else
            print '(A)', 'FAILURE: Solver did not converge'
        end if
    end if

    deallocate(u)
    deallocate(u_new)
    call dtl_context_destroy(ctx)

end program jacobi_1d
